### PR TITLE
atomic add

### PR DESCRIPTION
**Pull request by jiyang1011**
This patch is for atomic add feature which is a necessary component for split-k and stream-k algorithm.
1. limitation: the data type must be float / double / int [https://super-funicular-b7914353.pages.github.io/public/atomic-datatypes.html#data-types-in-atomic-operations]
2. the memory space is global 
3. Its `copy_traits` is same with `UniversalCopy`

### CODE REVIEW

File: include/cute/arch/copy_xe.hpp [hunk 1]
```diff
@@ -475,4 +475,26 @@ struct XE_2D_U32x8x16x1x1_ST_N
...
+struct XE_ATOMIC {
+  using SRegisters = S[1];
+  using DRegisters = D[1];
+
+  CUTE_STATIC_ASSERT(is_same_v<S, float> || is_same_v<S, double> || is_same_v<S, int>);
+
+  template<class S_, class D_>
+  CUTE_HOST_DEVICE static void
+  copy(S_ const& src, D_ & dst) {
+    #if defined(SYCL_INTEL_TARGET)
```
Comment (mehdi-goli) [inline include/cute/arch/copy_xe.hpp:489]:
```suggestion
    #if defined(__SYCL_DEVICE_ONLY__)
```

**Review by mehdi-goli — APPROVED** [review]

**Review by mehdi-goli — COMMENTED** [review]
Thanks for the PR. For the Split-k/StreamK, we need to have the flag memory and read/write from/to flag memory, and not loading/storing the actual data, similar to what [Nvidia did](https://github.com/codeplaysoftware/cutlass-fork/blob/64acac8f1043375d11dfbe93dca04a841f68e773/include/cutlass/barrier.h#L167).
I think it would be better to use SYCLCompat `Atomic_compare_and_exchange_strong` similar to  `atomicCAS` enabled for Nvida in the link I sent above Would you be able to look into the link and implement it like that? This is the function that the barrier manager in Nvidia implementation for Split/StreamK uses to update the flag memory.

So

**Review by mehdi-goli — COMMENTED** [review]
Also, since the title of your PR says atomic add (just to mention that we dont need atomic add for Splitk/Streamk), Nvidia implemented Atomic add in [the functional.h file](https://github.com/codeplaysoftware/cutlass-fork/blob/fae9b2e19f799ba7ffbadec70d3ce9bd1e2701ed/include/cutlass/functional.h#L577). Incase we need it for any other operations, it is better to follow that pattern and put it there.

**Comment by aacostadiaz** [comment]
> Thanks for the PR. For the Split-k/StreamK, we need to have the flag memory and read/write from/to flag memory, and not loading/storing the actual data, similar to what [Nvidia did](https://github.com/codeplaysoftware/cutlass-fork/blob/64acac8f1043375d11dfbe93dca04a841f68e773/include/cutlass/barrier.h#L167). I think it would be better to use SYCLCompat `Atomic_compare_and_exchange_strong` similar to `atomicCAS` enabled for Nvida in the link I sent above Would you be able to look into the link and implement it like that? This is the function that the barrier manager in Nvidia implementation for Split/StreamK uses to update the flag memory.

For reference, there is a SYCL implementation of `atomicCAS` in [gpu_generic.h](https://github.com/codeplaysoftware/cutlass-fork/blob/64acac8f1043375d11dfbe93dca04a841f68e773/include/cutlass/gpu_generics.h#L325). It just calls the syclcompat function @mehdi-goli mentioned.

**Comment by rolandschulz** [comment]
> just to mention that we dont need atomic add for Splitk/Streamk

What's the plan to do the reduction for each? 

Why would we implement barrier with CAS and not use the existing barrier?

**Comment by mehdi-goli** [comment]
> > just to mention that we dont need atomic add for Splitk/Streamk
> 
> What's the plan to do the reduction for each?
> 
> Why would we implement barrier with CAS and not use the existing barrier?

The Flag memory is the integer memory used to sync across the workgroup, hence we can avoid having separate kernel for reduction. With atomicCAS, you only update the boolean value for informing other workgroup that I am done with writing intermediate output. This usually get updated by workitem0 when all the threads within the workgroup to perform gemm once local barrier condition satisfied.  There is one memory flag per tile. For example, for the 2splitk tile algorithm (sharing the last 2 wave)  across 4 workgroup based on the problem size provided in [the paper](https://arxiv.org/pdf/2301.03598) (figure 3c) you need only one boolean flag memory the streamK tile. So the total would be 5 (NumStreamKTiles*NumSplits= 5x1).

For the reduction part, it is partial reduction and only the consumer workgroup need to check the flag, if the data is ready it can read it from global, sum it with its output which is currently in register and write it back to the final memory. 

Using atomic add for reducing the data help us to reduce the cost of temporary global memory usage. In this case we can use atomic add to reduce the global memory usage, but we don't need to implement it in the copy function. Nvidia put the implementation of atomic add [here](https://github.com/codeplaysoftware/cutlass-fork/blob/fae9b2e19f799ba7ffbadec70d3ce9bd1e2701ed/include/cutlass/functional.h#L577) and it uses for [reduction inside the algorithm](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/include/cutlass/block_striped.h#L221) If we go for copy approach we will create a deviation from Nvidia approaches which we should properly document it and say why we need to put it in the copy.

**Comment by jiyang1011** [comment]
> Also, since the title of your PR says atomic add (just to mention that we dont need atomic add for Splitk/Streamk), Nvidia implemented Atomic add in [the functional.h file](https://github.com/codeplaysoftware/cutlass-fork/blob/fae9b2e19f799ba7ffbadec70d3ce9bd1e2701ed/include/cutlass/functional.h#L577). Incase we need it for any other operations, it is better to follow that pattern and put it there.

Hi mehdi, I checked the code of Xetla. I think you are right, atomic add is used for sync signal. But I have a question that `atomicCAS` is under cutlass 2.0, whether a customized `CollectiveMMA`  should be re-written for stream-k algorithm and some necessary components such as memory fence, signal sync as utils?

**Comment by jiyang1011** [comment]
> > Thanks for the PR. For the Split-k/StreamK, we need to have the flag memory and read/write from/to flag memory, and not loading/storing the actual data, similar to what [Nvidia did](https://github.com/codeplaysoftware/cutlass-fork/blob/64acac8f1043375d11dfbe93dca04a841f68e773/include/cutlass/barrier.h#L167). I think it would be better to use SYCLCompat `Atomic_compare_and_exchange_strong` similar to `atomicCAS` enabled for Nvida in the link I sent above Would you be able to look into the link and implement it like that? This is the function that the barrier manager in Nvidia implementation for Split/StreamK uses to update the flag memory.
> 
> For reference, there is a SYCL implementation of `atomicCAS` in [gpu_generic.h](https://github.com/codeplaysoftware/cutlass-fork/blob/64acac8f1043375d11dfbe93dca04a841f68e773/include/cutlass/gpu_generics.h#L325). It just calls the syclcompat function @mehdi-goli mentioned.
(https://github.com/codeplaysoftware/cutlass-fork/blob/64acac8f1043375d11dfbe93dca04a841f68e773/include/cutlass/gpu_generics.h#L325). seems missing` return` ?

**Comment by jiyang1011** [comment]
> > Also, since the title of your PR says atomic add (just to mention that we dont need atomic add for Splitk/Streamk), Nvidia implemented Atomic add in [the functional.h file](https://github.com/codeplaysoftware/cutlass-fork/blob/fae9b2e19f799ba7ffbadec70d3ce9bd1e2701ed/include/cutlass/functional.h#L577). Incase we need it for any other operations, it is better to follow that pattern and put it there.
> 
> Hi mehdi, I checked the code of Xetla. I think you are right, atomic add is used for sync signal. But I have a question that `atomicCAS` is under cutlass 2.0, whether a customized `CollectiveMMA` should be re-written for stream-k algorithm and some necessary components such as memory fence, signal sync as utils?

[](https://github.com/intel/xetla/blob/main/include/group/epilogue/impl/stream_k_op_xe.hpp#L134) this store action used atomic add for reduction.

**Comment by rolandschulz** [comment]
I don't think it is missing a `return`. The return is outside the `#endif` at https://github.com/codeplaysoftware/cutlass-fork/blob/64acac8f1043375d11dfbe93dca04a841f68e773/include/cutlass/gpu_generics.h#L329 . 

But I do think the return is incorrect. [atomicCAS()](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomiccas) should return the old value. Whereas it currently always returns 0. I don't think it is possible to correctly implement `atomicCAS` with `atomic_ref` (sycl or std) because it doesn't have an operation which compares and returns the old value atomicly. All usage (https://github.com/search?q=repo%3ANVIDIA%2Fcutlass+atomiccas&type=code) of the return value just compare it to desired. I suggest we replace `atomicCAS` with `atomic_compare_exchange` which directly maps to std/sycl. And for CUDA it maps to `return atomicCAS(address, expected, desired)==expected`.

**Comment by aacostadiaz** [comment]
Yes, it is missing the return. Good catch! It should be something like: 

```c++
CUTLASS_DEVICE int atomicCAS(int *address, int compare, int val) {
#if defined(__SYCL_DEVICE_ONLY__)
  return syclcompat::atomic_compare_exchange_strong(address, compare, val);
#else
  CUTLASS_NOT_IMPLEMENTED()
  return 0;
#endif
}
```

`syclcompat::atomic_compare_exchange_strong` returns the value in `address` before the call.

**Comment by aacostadiaz** [comment]
> > Also, since the title of your PR says atomic add (just to mention that we dont need atomic add for Splitk/Streamk), Nvidia implemented Atomic add in [the functional.h file](https://github.com/codeplaysoftware/cutlass-fork/blob/fae9b2e19f799ba7ffbadec70d3ce9bd1e2701ed/include/cutlass/functional.h#L577). Incase we need it for any other operations, it is better to follow that pattern and put it there.
> 
> Hi mehdi, I checked the code of Xetla. I think you are right, atomic add is used for sync signal. But I have a question that `atomicCAS` is under cutlass 2.0, whether a customized `CollectiveMMA` should be re-written for stream-k algorithm and some necessary components such as memory fence, signal sync as utils?

`atomicCAS` is a Cuda function used in CUTLASS. It is used in both CUTLASS 2 and 3.

Regarding `atomic_add`, could you please check if the CUTLASS implementation of `struct atomic_add` in [functional.h](https://github.com/NVIDIA/cutlass/blob/f7b19de32c5d1f3cedfc735c2849f12b537522ee/include/cutlass/functional.h#L773) covers the use case you are trying to address with this PR? Nvidia uses that struct as a base component to implement other functionalities (such as reduction). If possible, we would like to follow the same path to avoid reimplementing everything for Intel.

**Comment by muhammad-tanvir-1211** [comment]
> > Also, since the title of your PR says atomic add (just to mention that we dont need atomic add for Splitk/Streamk), Nvidia implemented Atomic add in [the functional.h file](https://github.com/codeplaysoftware/cutlass-fork/blob/fae9b2e19f799ba7ffbadec70d3ce9bd1e2701ed/include/cutlass/functional.h#L577). Incase we need it for any other operations, it is better to follow that pattern and put it there.
> 
> Hi mehdi, I checked the code of Xetla. I think you are right, atomic add is used for sync signal. But I have a question that `atomicCAS` is under cutlass 2.0, whether a customized `CollectiveMMA` should be re-written for stream-k algorithm and some necessary components such as memory fence, signal sync as utils?

Hi @jiyang1011, I believe we can reuse the existing the `CollectiveMma` atom (with minor modifications) for splitK/streamK GEMM as well. We would definitely need to implement a specialized pipeline for splitK/streamK similar to what NVIDIA does (e.g. [sm90 streamK pipeline](https://github.com/codeplaysoftware/cutlass-fork/blob/17ce5ff5a0d4743cae52c79588570ec18e9a3ae0/include/cutlass/gemm/kernel/sm90_gemm_warpspecialized_cooperative.hpp#L248))

**Comment by rolandschulz** [comment]
> syclcompat::atomic_compare_exchange_strong returns the value in address before the call.

It doesn't: https://github.com/intel/llvm/issues/15155 . Even if the bug is fixed, using this API isn't very efficient. It requires and extra unnecessary load and comparison. All usage of the call just care about whether `old==expected`. Which is the boolean return value of (std/sycl::)`atomic_ref::atomic_compare_exchange`. But because the API returns the old value rather than the boolean, the calling code has to do an extra comparison and the implementation of `syclcompat::atomic_compare_exchange_strong` needs to do an extra load and extra comparison.

**Comment by aacostadiaz** [comment]
> > syclcompat::atomic_compare_exchange_strong returns the value in address before the call.
> 
> It doesn't: [intel/llvm#15155](https://github.com/intel/llvm/issues/15155) . Even if the bug is fixed, using this API isn't very efficient. It requires and extra unnecessary load and comparison. All usage of the call just care about whether `old==expected`. Which is the boolean return value of (std/sycl::)`atomic_ref::atomic_compare_exchange`. But because the API returns the old value rather than the boolean, the calling code has to do an extra comparison and the implementation of `syclcompat::atomic_compare_exchange_strong` needs to do an extra load and extra comparison.

I think there is no bug. `atm.compare_exchange_strong(expected, ...)` will handle the comparison and assign the value in `address` to `expected` if needed. So, after calling `atm.compare_exchange_strong`, `expected` will contain either the expected value or the old value, depending on the comparison.

https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#table.atomics.functions
image

Regarding the API being inefficient, I think you are right. We can look into `bool syclcompat::compare_exchange_strong` or just use the SYCL function directly.

**Pull request closed without merging**